File: bindings/imgui_bundle/demos_cpp/demos_nanovg/demo_nanovg_heart.cpp
#include "hello_imgui/hello_imgui.h"

#include "imgui.h"
#include "nanovg.h"
#include "nvg_imgui/nvg_imgui.h"
#include <math.h>


struct DrawingState
{
    ImVec4 HeartColor = ImVec4(1.f, 0.f, 0.f, 1.f);
};

DrawingState gDrawingState;


void DrawHeart(NVGcontext* vg, ImVec2 center, float size, float t)
{
    float x = center.x, y = center.y;
    float scale = sinf(t * 10) * 0.1f + 1.0f;  // Oscillating scale for the beating effect
    float scaledSize = size * scale;

    nvgSave(vg);

    // Change coord so that we draw in coords between (-1, 1) and (-1, 1)
    nvgTranslate(vg, x, y);
    nvgScale(vg, scaledSize, -scaledSize); // y points up

    nvgBeginPath(vg);
    {
        nvgMoveTo(vg, 0.f, 0.4f);
        nvgBezierTo(vg, 0.f, 0.5f,     0.1f, 1.f,     0.5f, 1.f);
        nvgBezierTo(vg, 0.9f, 1.f,     1.f, 0.7f,     1.f, 0.4f);
        nvgBezierTo(vg, 1.f, 0.2f,     0.75f, -0.2f,  0.5f, -0.4f);
        nvgBezierTo(vg, 0.2f, -0.65f,  0.f, -0.8f,    0.f, -1.f);

        nvgBezierTo(vg, 0.f, -0.8f,    -0.2f, -0.65f,  -0.5f, -0.4f);
        nvgBezierTo(vg, -0.75f, -0.2f, -1.f, 0.2f,     -1.f, 0.4f);
        nvgBezierTo(vg, -1.f, 0.7f,    -0.9f, 1.f,     -0.5f, 1.f);
        nvgBezierTo(vg, -0.1f, 1.f,    0.f, 0.5f,      0.f, 0.4f);
    }

    // Create gradient from top to bottom
    ImVec4 color = gDrawingState.HeartColor;
    auto color_full = nvgRGBAf(color.x, color.y, color.z, color.w);
    auto color_transparent = nvgRGBAf(color.x, color.y, color.z, 0.2f);
    NVGpaint paint = nvgLinearGradient(vg, 0.f, 1.f, 0.f, -1.f, color_full, color_transparent);
    nvgFillPaint(vg, paint);
    nvgFill(vg);

    nvgStrokeColor(vg, nvgRGBA(0, 0, 255, 255));
    nvgStrokeWidth(vg, 0.05f);
    nvgStroke(vg);

    nvgRestore(vg);
}


void DrawNanoVgLabel(NVGcontext* vg, float width, float height)
{
    // The font used to write "NanoVG": it should be loaded only once
    static int fontId = -1;
    if (fontId == -1)
    {
        // Load the font
        #ifndef __ANDROID__
        auto fontPath = HelloImGui::AssetFileFullPath("fonts/Roboto/Roboto-Bold.ttf");
        fontId = nvgCreateFont(vg, "roboto", fontPath.c_str());
        #else
        // On Android, assets are compressed inside the apk, so we need to load them differently
        auto assetData = HelloImGui::LoadAssetFileData("fonts/Roboto/Roboto-Bold.ttf");
        fontId = nvgCreateFontMem(vg, "roboto", (unsigned char*)assetData.data, assetData.dataSize, 0);
        HelloImGui::FreeAssetFileData(&assetData);
        #endif
        if (fontId == -1) {
            fprintf(stderr, "Could not add font.\n");
            return; // Exit if the font cannot be added.
        }
    }

    nvgBeginPath(vg);
    nvgFontSize(vg, 128.f);
    nvgFontFaceId(vg, fontId);
    nvgTextAlign(vg, NVG_ALIGN_CENTER | NVG_ALIGN_MIDDLE);
    nvgSave(vg);
    nvgRotate(vg, -0.1f);
    nvgFillColor(vg, nvgRGBA(255, 100, 100, 255));
    nvgText(vg, 0.5f * width, 0.9f * height, "NanoVG", nullptr);
    nvgRestore(vg);
}


void DrawScene(NVGcontext* vg, float width, float height)
{
    nvgSave(vg);

    // Draw a white background
    nvgBeginPath(vg);
    nvgRect(vg, 0, 0, width, height);
    nvgFillColor(vg, nvgRGBA(255, 255, 255, 255));
    nvgFill(vg);

    // Draw NanoVG
    DrawNanoVgLabel(vg, width, height);

    // Draw a heart
    DrawHeart(vg, {width / 2.f, height / 2.f}, width * 0.2f, ImGui::GetTime());

    nvgRestore(vg);
};


struct AppStateNvgHeart
{
    // Our NanoVG context
    NVGcontext *vg = nullptr;

    // A framebuffer, which will be used as a texture for the ImGui rendering
    std::unique_ptr<NvgImgui::NvgFramebuffer> nvgFramebuffer;

    void Init()
    {
        // Instantiate the NanoVG context
        vg = NvgImgui::CreateNvgContext_HelloImGui(NvgImgui::NVG_ANTIALIAS | NvgImgui::NVG_STENCIL_STROKES);

        // Create a framebuffer
        int nvgImageFlags = 0; //NVG_IMAGE_FLIPY | NVG_IMAGE_PREMULTIPLIED;
        nvgFramebuffer = std::make_unique<NvgImgui::NvgFramebuffer>(vg, 1000, 600, nvgImageFlags);
    }

    void Release()
    {
        nvgFramebuffer.reset();
        NvgImgui::DeleteNvgContext_HelloImGui(vg);
    }
};


int main(int, char**)
{
    AppStateNvgHeart appState;

    HelloImGui::RunnerParams runnerParams;

    runnerParams.callbacks.EnqueuePostInit([&]() { appState.Init(); });
    runnerParams.callbacks.EnqueueBeforeExit([&]() { appState.Release(); });


    auto gui = [&]()
    {
        ImGui::Text("This image below is rendered by NanoVG, via a framebuffer.");
        // Render our drawing to a framebuffer, and use it as a texture for ImGui
        NvgImgui::RenderNvgToFrameBuffer(appState.vg, *appState.nvgFramebuffer, DrawScene);
        ImGui::Image(appState.nvgFramebuffer->TextureId, HelloImGui::EmToVec2(50.f, 30.f));

        ImGui::ColorEdit4("Heart color", &gDrawingState.HeartColor.x);
    };
    runnerParams.callbacks.ShowGui = gui;
    runnerParams.fpsIdling.enableIdling = false;

    HelloImGui::Run(runnerParams);
}

File: bindings/imgui_bundle/demos_python/demos_nanovg/demo_nanovg_heart.py
from imgui_bundle import hello_imgui, imgui, nanovg as nvg, ImVec2
import math
from typing import List

nvg_imgui = nvg.nvg_imgui


class DrawingState:
    heart_color: List[float]

    def __init__(self):
        self.heart_color = [1.0, 0.0, 0.0, 1.0]

gDrawingState = DrawingState()


def draw_heart(vg: nvg.Context, center: ImVec2, size: float, t: float):
    x = center.x
    y = center.y
    scale = math.sin(t * 10) * 0.1 + 1.0  # Oscillating scale for the beating effect
    scaled_size = size * scale

    nvg.save(vg)

    # Change coord so that we draw in coords between (-1, 1) and (-1, 1)
    nvg.translate(vg, x, y)
    nvg.scale(vg, scaled_size, -scaled_size)  # y points up

    nvg.begin_path(vg)
    nvg.move_to(vg, 0.0, 0.4)
    nvg.bezier_to(vg, 0.0, 0.5, 0.1, 1.0, 0.5, 1.0)
    nvg.bezier_to(vg, 0.9, 1.0, 1.0, 0.7, 1.0, 0.4)
    nvg.bezier_to(vg, 1.0, 0.2, 0.75, -0.2, 0.5, -0.4)
    nvg.bezier_to(vg, 0.2, -0.65, 0.0, -0.8, 0.0, -1.0)

    nvg.bezier_to(vg, 0.0, -0.8, -0.2, -0.65, -0.5, -0.4)
    nvg.bezier_to(vg, -0.75, -0.2, -1.0, 0.2, -1.0, 0.4)
    nvg.bezier_to(vg, -1.0, 0.7, -0.9, 1.0, -0.5, 1.0)
    nvg.bezier_to(vg, -0.1, 1.0, 0.0, 0.5, 0.0, 0.4)

    # Create gradient from top to bottom
    color = gDrawingState.heart_color
    color_full = nvg.rgba_f(color[0], color[1], color[2], color[3])
    color_transparent = nvg.rgba_f(color[0], color[1], color[2], 0.2)
    paint = nvg.linear_gradient(vg, 0.0, 1.0, 0.0, -1.0, color_full, color_transparent)
    nvg.fill_paint(vg, paint)
    nvg.fill(vg)

    nvg.stroke_color(vg, nvg.rgba(0, 0, 255, 255))
    nvg.stroke_width(vg, 0.05)
    nvg.stroke(vg)

    nvg.restore(vg)


def draw_nano_vg_label(vg: nvg.Context, width: float, height: float):
    # The font used to write "NanoVG": it should be loaded only once
    static = draw_nano_vg_label
    if not hasattr(static, "static_font_id"):
        # Load the font
        font_path = hello_imgui.asset_file_full_path("fonts/Roboto/Roboto-Bold.ttf")
        static.font_id = nvg.create_font(vg, "roboto", font_path)
        if static.font_id == -1:
            print("Could not add font.")
            return  # Exit if the font cannot be added.

    nvg.begin_path(vg)
    nvg.font_size(vg, 128.0)
    nvg.font_face_id(vg, static.font_id)
    nvg.text_align(vg, nvg.Align.align_center.value | nvg.Align.align_middle.value)
    nvg.save(vg)
    nvg.rotate(vg, -0.1)
    nvg.fill_color(vg, nvg.rgba(255, 100, 100, 255))
    nvg.text(vg, 0.5 * width, 0.9 * height, "NanoVG")
    nvg.restore(vg)


def draw_scene(vg: nvg.Context, width: float, height: float):
    nvg.save(vg)

    # Draw a white background
    nvg.begin_path(vg)
    nvg.rect(vg, 0, 0, width, height)
    nvg.fill_color(vg, nvg.rgba(255, 255, 255, 255))
    nvg.fill(vg)

    # Draw NanoVG
    draw_nano_vg_label(vg, width, height)

    # Draw a heart
    draw_heart(vg, ImVec2(width / 2.0, height / 2.0), width * 0.2, imgui.get_time())

    nvg.restore(vg)



class AppStateNvgHeart:
    vg: nvg.Context = None
    nvg_framebuffer: nvg_imgui.NvgFramebuffer = None

    def __init__(self):
        # Our NanoVG context
        self.vg = None

        # A framebuffer, which will be used as a texture for the ImGui rendering
        self.nvg_framebuffer = None

    def init(self):
        # Instantiate the NanoVG context
        self.vg = nvg_imgui.create_nvg_context_hello_imgui(nvg_imgui.NvgCreateFlags.antialias | nvg_imgui.NvgCreateFlags.stencil_strokes)

        # Create a framebuffer
        nvg_image_flags = 0  # NVG_IMAGE_FLIPY | NVG_IMAGE_PREMULTIPLIED;
        self.nvg_framebuffer = nvg_imgui.NvgFramebuffer(self.vg, 1000, 600, nvg_image_flags)

    def release(self):
        self.nvg_framebuffer = None
        nvg_imgui.delete_nvg_context_hello_imgui(self.vg)


def main():
    app_state = AppStateNvgHeart()

    runner_params = hello_imgui.RunnerParams()

    runner_params.callbacks.enqueue_post_init(lambda: app_state.init())
    runner_params.callbacks.enqueue_before_exit(lambda: app_state.release())

    def gui():
        imgui.text("This image below is rendered by NanoVG, via a framebuffer.")

        # Render our drawing to a framebuffer, and use it as a texture for ImGui
        nvg_imgui.render_nvg_to_frame_buffer(app_state.vg, app_state.nvg_framebuffer, draw_scene)
        imgui.image(app_state.nvg_framebuffer.texture_id, hello_imgui.em_to_vec2(50, 30))

        _, gDrawingState.heart_color = imgui.color_edit4("Heart color", gDrawingState.heart_color)

    runner_params.callbacks.show_gui = gui
    runner_params.fps_idling.enable_idling = False

    hello_imgui.run(runner_params)


if __name__ == "__main__":
    main()

